xdsl/transforms/experimental/stencil_global_to_local.py

```diff
@@ -0,0 +1,491 @@
+            ),
+            neighbor=-1,
+        )
+        # lower halo exchange:
+        yield HaloExchangeDef(
+            offset=(
+                dims.core_start(dims.DIM_X),
+                dims.core_end(dims.DIM_Y),
+            ),
+            size=(
+                dims.core_size(dims.DIM_X),
+                dims.halo_size(dims.DIM_Y),
+            ),
+            source_offset=(
+                0,
+                -dims.halo_size(dims.DIM_Y),
+            ),
+            neighbor=1,
+        )
+
+
+@dataclass
+class ChangeStoreOpSizes(RewritePattern):
+    strategy: SlicingStrategy
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: stencil.StoreOp, rewriter: PatternRewriter,
+                          /):
+        assert all(integer_attr.value.data == 0
+                   for integer_attr in op.lb.array.data), "lb must be 0"
+        shape: tuple[int, ...] = tuple(
+            (integer_attr.value.data for integer_attr in op.ub.array.data))
+        new_shape = self.strategy.calc_resize(shape)
+        op.ub = stencil.IndexAttr.get(*new_shape)
+
+
+@dataclass
+class AddHaloExchangeOps(RewritePattern):
+    """
+    This rewrite adds a `stencil.halo_exchange` before each `stencil.load` op
+    """
+    strategy: SlicingStrategy
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: stencil.LoadOp, rewriter: PatternRewriter,
+                          /):
+        swap_op = stencil.HaloSwapOp.get(op.res)
+        rewriter.insert_op_after_matched_op(swap_op)
+
+
+def global_stencil_to_local_stencil_2d_horizontal(ctx: MLContext | None,
+                                                  module: builtin.ModuleOp):
+    strategy = HorizontalSlices2D(2)
+
+    gpra = GreedyRewritePatternApplier(
+        [ChangeStoreOpSizes(strategy),
+         AddHaloExchangeOps(strategy)])
+
+    PatternRewriteWalker(gpra, apply_recursively=False).rewrite_module(module)
+
+
+@dataclass
+class LowerHaloExchangeToMpi(RewritePattern):
+    strategy: SlicingStrategy
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: stencil.HaloSwapOp,
+                          rewriter: PatternRewriter, /):
+        exchanges = list(
+            self.strategy.halo_exchange_defs(DimsHelper.from_halo_swap_op(op)))
+        assert isa(op.input_stencil.typ, memref.MemRefType[Attribute])
+        rewriter.replace_matched_op(
+            list(
+                generate_mpi_calls_for(
+                    op.input_stencil,
+                    exchanges,
+                    op.input_stencil.typ.element_type,
+                    self.strategy,
+                )),
+            [],
+        )
+
+
+def generate_mpi_calls_for(source: SSAValue, exchanges: list[HaloExchangeDef],
+                           dtype: Attribute,
+                           strat: SlicingStrategy) -> Iterable[Operation]:
+    # allocate request array
+    req_cnt = arith.Constant.from_int_and_width(
+        len(exchanges) * 2, builtin.i32)
```

Comment:
Can you use a parameter var for 2?

Reply:
I added a comment to explain this further